packages/@n8n/db/src/repositories/workflow-dependency.repository.ts

```diff
@@ -0,0 +1,181 @@
+				// Perform the update using queryRunner.manager
+				const result = await this.executeUpdate(workflowId, dependencies, queryRunner.manager);
+
+				await queryRunner.query('COMMIT');
+				return result;
+			} catch (error) {
+				await queryRunner.query('ROLLBACK');
+				throw error;
+			}
+		} finally {
+			await queryRunner.release();
+		}
+	}
+
+	private async executeUpdate(
+		workflowId: string,
+		dependencies: WorkflowDependencies,
+		tx: EntityManager,
+	): Promise<boolean> {
+		const deleteResult = await tx.delete(WorkflowDependency, {
+			workflowId,
+			workflowVersionId: LessThan(dependencies.workflowVersionId),
+		});
+
+		// If we deleted something, the incoming version is newer - proceed with insert
+		if (deleteResult.affected && deleteResult.affected > 0) {
+			const entities = dependencies.dependencies.map((dep) => this.create(dep));
+			// NOTE: we cast to any[] because TypeORM doesn't like the JSON column.
+			// eslint-disable-next-line @typescript-eslint/no-explicit-any
+			await tx.insert(WorkflowDependency, entities as any[]);
+			return true;
+		}
+
+		// Nothing was deleted - either no existing data, or existing data is newer/same version
+		// Check if any dependencies exist for this workflow. We lock for update to avoid a race
+		// when two processes try to insert dependencies for the same workflow at the same time.
+		const hasData = await this.hasExistingData(workflowId, tx);
+
+		if (!hasData) {
+			// There's no existing data, so we can safely insert the new dependencies.
+			const entities = dependencies.dependencies.map((dep) => this.create(dep));
+			// NOTE: we cast to any[] because TypeORM doesn't like the JSON column.
+			// eslint-disable-next-line @typescript-eslint/no-explicit-any
+			await tx.insert(WorkflowDependency, entities as any[]);
+			return true;
+		}
+
+		// Existing data has same or newer version - skip update
+		return false;
+	}
+
+	/**
+	 * Remove all dependencies for a given workflow.
+	 *
+	 * NOTE: there's a possible race in case of an update and delete happening concurrently.
+	 * The delete could be reflected in the database, but the update could be reflected in the index.
+	 * To prevent this we would need to implement some tombstone mechanism. However, since we don't
+	 * do this for the workflow itself, it would be inconsistent to do it only for the dependencies.
+	 * The chance of this happening in practice is also very low.
+	 *
+	 * @param workflowId The ID of the workflow
+	 * @returns Whether any dependencies were removed
+	 */
+	async removeDependenciesForWorkflow(workflowId: string): Promise<boolean> {
+		return await this.manager.transaction(async (tx) => {
+			const deleteResult = await tx.delete(WorkflowDependency, { workflowId });
+
+			return (
+				deleteResult.affected !== undefined &&
+				deleteResult.affected !== null &&
+				deleteResult.affected > 0
+			);
+		});
+	}
+
+	private async hasExistingData(workflowId: string, tx: EntityManager): Promise<boolean> {
```

Comment:
The name of this is a bit confusing as it checks the existence AND also does locking